### PR TITLE
[SPARK-32207][SQL] Support 'F'-suffixed Float Literals

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

In this PR, I suppose we support 'f'-suffixed float literal, e.g. `select 1.1f`

### Why are the changes needed?

a very common feature across platforms, checked with pg, presto, hive, MySQL...

### Does this PR introduce _any_ user-facing change?

yes, 

`select 1.1f` results float value 1.1 instead of throwing AnlaysisExceptiion`Can't extract value from 1: need struct type but got int;`

### How was this patch tested?

add unit tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @HyukjinKwon thanks for reviewing this in advance.

**Comment by cloud-fan** [comment]
I don't know why it's not supported before, and looks fine to me. cc @maropu @viirya  @bart-samwel

**Comment by bart-samwel** [comment]
SGTM!

**Comment by HyukjinKwon** [comment]
+1

**Comment by maropu** [comment]
Oh, I see. Looks okay to add this.

**Review by wangyum — COMMENTED** [review]
Hive 2.3 and PostgreSQL seem to be aliases:
```sql
hive> CREATE TABLE test_float1 as SELECT 1.0F, 2.0D;
hive> DESC test_float1;
OK
f                   	decimal(1,0)
_c1                 	double
Time taken: 0.763 seconds, Fetched: 2 row(s)
```

```sql
postgres=# CREATE TABLE test_float1 as SELECT 1.0F, 2.0D;
postgres=# \d test_float1;
            Table "public.test_float1"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 f      | numeric |           |          |
 d      | numeric |           |          |
```

**Comment by bart-samwel** [comment]
> Hive 2.3 and PostgreSQL seem to be aliases:
> 
> ```sql
> hive> CREATE TABLE test_float1 as SELECT 1.0F, 2.0D;
> hive> DESC test_float1;
> OK
> f                   	decimal(1,0)
> _c1                 	double
> Time taken: 0.763 seconds, Fetched: 2 row(s)
> ```
> 
> ```sql
> postgres=# CREATE TABLE test_float1 as SELECT 1.0F, 2.0D;
> postgres=# \d test_float1;
>             Table "public.test_float1"
>  Column |  Type   | Collation | Nullable | Default
> --------+---------+-----------+----------+---------
>  f      | numeric |           |          |
>  d      | numeric |           |          |
> ```

 I wonder what postgresql does with `1.0e2` -- does it treat it like 1.0 with alias e2 or like 1.0 * 10^2? And `1.0e` probably is an alias then. These kinds of discontinuities are terrible and we should stay far away from them...

**Comment by wangyum** [comment]
```sql
postgres=# CREATE TABLE test_float2 as SELECT 1.0F, 2.0D, 1.0e2, 1.0e;
SELECT 1
postgres=# \d test_float2;
             Table "public.test_float2"
  Column  |  Type   | Collation | Nullable | Default
----------+---------+-----------+----------+---------
 f        | numeric |           |          |
 d        | numeric |           |          |
 ?column? | numeric |           |          |
 e        | numeric |           |          |

postgres=# select * from test_float2;
  f  |  d  | ?column? |  e
-----+-----+----------+-----
 1.0 | 2.0 |      100 | 1.0
(1 row)
```

**Comment by viirya** [comment]
I noticed this too when I played with other PR. I thought it was by purpose. If it is not, looks okay to add it.

**Comment by cloud-fan** [comment]
retest this please

**Comment by maropu** [comment]
Could you update the doc, too? https://github.com/apache/spark/blob/master/docs/sql-ref-literals.md#numeric-literal
cc: @huaxingao

**Comment by yaooqinn** [comment]
documentation updated, thanks for reminding me @maropu

File: docs/sql-ref-literals.md [hunk 1]
```diff
@@ -239,6 +244,10 @@ E [ + | - ] digit [ ... ]
 
     Case insensitive, indicates `DOUBLE`, which is an 8-byte double-precision floating point number.
 
+* **D**
```
Comment (cloud-fan) [inline docs/sql-ref-literals.md:247]: F

**Review by maropu — APPROVED** [review]
LGTM except for the @cloud-fan comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -2030,6 +2030,12 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
       Long.MinValue, Long.MaxValue, LongType.simpleString)(_.toLong)
   }
 
+  override def visitFloatLiteral(ctx: FloatLiteralContext): Literal = {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2033]: Could you add a function description like the other functions around here, @yaooqinn ?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (except @cloud-fan 's comment and mine)

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2036]: done, thanks!

**Comment by wangyum** [comment]
Maybe we should convert suffix to alias. Please see these examples for more details:

Presto:
image

MySQL 5.7.26:
image

MySQL 8.0.16:
image

MariaDB 10.4.6:
image

PostgreSQL 11.3:
image

Hive 3.1.2:
image

**Comment by yaooqinn** [comment]
> Maybe we should convert suffix to alias. Please see these examples for more details:

If the digits part is integral, e.g. `2f`, `2a`, these systems vary from each other..., not always alias. on the other hand, inside spark, we have different error msg for `2.1f`(Can't extract value from 1: need struct type but got int;) and `2f`(cannot resolve '`2f`' given input columns: []), but `2d` and `2.1d` are valid inputs. IMHO, it's better to keep the consistency of our
own first.

**Comment by wangyum** [comment]
Is there a system support `F`-suffixed as float Literal?

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
Seems like most databases don't support the `D` suffix. Spark can't drop the `D` support now, I think it makes more sense to be consistent and support `F` as well.

**Comment by maropu** [comment]
Yea, the @cloud-fan comment above sounds fine to me.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
Thank you, @yaooqinn and all. Merged to master.

**Comment by yaooqinn** [comment]
Thanks for merging @dongjoon-hyun and all for the help

**Pull request closed without merging**